engine/linksystem.go

```diff
@@ -151,56 +154,39 @@ func (e *Engine) mkLinkSystem() ipld.LinkSystem {
 // This function takes a linksystem for persistence along with the channels
 // from a callback, and generates the linked list structure. It also supports
 // configuring the number of entries per chunk in the list.
-func generateChunks(lsys ipld.LinkSystem, chmhs <-chan mh.Multihash, cherr <-chan error, numEntries int) (ipld.Link, error) {
-	mhs := make([]mh.Multihash, 0, numEntries)
+func generateChunks(lsys ipld.LinkSystem, mhIter provider.MultihashIterator, maxChunkSize int) (ipld.Link, error) {
+	mhs := make([]mh.Multihash, 0, maxChunkSize)
 	var chunkLnk ipld.Link
-	var err error
-
-	var cidCount, chunkCount int
-
-	// For each Cid from callback.
-eachCid:
+	var totalMhCount, chunkCount int
 	for {
-		select {
-		case ec, ok := <-chmhs:
-			if !ok {
-				break eachCid
-			}
-			mhs = append(mhs, ec)
-		case err = <-cherr:
-			if err != nil {
-				// If something in error channel return error
-				return nil, err
+		next, err := mhIter.Next()
+		if err != nil {
+			if err == io.EOF {
+				if len(mhs) != 0 {
+					chunkLnk, _, err = schema.NewLinkedListOfMhs(lsys, mhs, chunkLnk)
+					if err != nil {
+						return nil, err
+					}
+					chunkCount++
+				}
+				break
 			}
-			break eachCid
+			return nil, err
 		}
+		mhs = append(mhs, next)
+		totalMhCount++
 
-		// Create chunk of cids and link to previous chunk
-		if len(mhs) == numEntries {
-			cidCount += len(mhs)
-			chunkCount++
-			// Create the chunk and restart variables
+		if len(mhs) == maxChunkSize {
```

Comment:
nit: I know this code is older, but it would be nicer to write it defensively with a `>=`, in case our code is buggy and we increment by more than one or partially skip an iteration.